### PR TITLE
Raise an exception when there is a course student with a negative xp when the semester is closed; Closes #1159

### LINKED ISSUE #1159: Closing a semester when student has negative XP
coursestudent.final_xp and final_grade can not be negative. However, if a student has negative XP (for example, by adding a negative xp_adjustment on the studentcourse) when closing a semester, the app will try to save that negative value into final_xp causing an error.

This needs to be handled somehow.  Ideally, with a warning that students can't finish a course with negative XP (similar to how it won't let you close the semester if there are still submissions to approve/reject)

Alternately (easier) it just sets a floor of 0 XP before saving final_xp adjustment

### PR DESCRIPTION
**Pull request by jefcolbi**

### CODE REVIEW

File: src/courses/models.py [hunk 1]
```diff
@@ -470,6 +474,7 @@ def calc_semester_grades(self, semester):
         coursestudents = self.get_queryset().get_semester(semester)
         for coursestudent in coursestudents:
             coursestudent.final_xp = coursestudent.user.profile.xp_per_course()
+            assert coursestudent.final_xp >= 0, f"{coursestudent.user.get_full_name()} has a negative XP. Fix it before closing the semester"
```
Comment (tylerecouture) [inline src/courses/models.py:477]: How is this going to inform the user?

File: src/courses/tests/test_models.py [hunk 2]
```diff
@@ -340,6 +340,13 @@ def test_xp_per_day_ave(self, days_so_far):
         xp_per_day = self.course_student.xp_per_day_ave()
         self.assertEqual(xp_per_day, 0)
 
+    @patch('profile_manager.models.Profile.xp_per_course')
+    def test_student_with_negative_xp(self, xp_per_course):
+        """Test if an assertion error is raised when there is a student with negative xp"""
+        xp_per_course.return_value = -10
+        self.assertRaises(AssertionError, CourseStudent.objects.calc_semester_grades,
+                          Semester.objects.get_current())
+
```
Comment (tylerecouture) [inline src/courses/tests/test_models.py:349]: Excellent!  Can you also test that the View that calls this method returns successfully in cases where there is negative XP, and that the warning is displayed on the page?  Currently it returns 403 or 404.

**Review by tylerecouture — CHANGES_REQUESTED** [review]
This looks great.  A couple requests, see the review comments. Thank you!

File: src/courses/models.py (hunk 1 above)
Comment (jefcolbi) [reply src/courses/models.py:477]: The method complete_active_semester() catches this exception and return Semester.STUDENTS_WITH_NEGATIVE_XP which is in turn handled by the view.

Comment (tylerecouture) [reply src/courses/models.py:477]: Ok, let's remove the assert, we don't need that in our production code.

File: src/courses/tests/test_views.py [hunk 3]
```diff
@@ -664,6 +666,29 @@ def test_SemesterUpdate__delete_data__view(self):
 
         self.assertFalse(ExcludedDate.objects.exists())
 
+    @patch('profile_manager.models.Profile.xp_per_course')
+    def test_SemesterDCloses__student_with_negative_xp__view(self, xp_per_course):
```
Comment (tylerecouture) [inline src/courses/tests/test_views.py:670]: Needs docstring

File: src/courses/tests/test_views.py [hunk 4]
```diff
@@ -664,6 +666,29 @@ def test_SemesterUpdate__delete_data__view(self):
...
+
+        semester = Semester.objects.first()
+        student = baker.make(User)
+        course = baker.make(Course)
+        baker.make(CourseStudent, user=student, course=course, semester=semester)
+
+        response = self.client.post(reverse('courses:end_active_semester'))
+        self.assertWarningMessage(response)
+        message = self.get_message_list(response)[0]
+        self.assertIn('There are some students with negative XP', str(message))
```
Comment (tylerecouture) [inline src/courses/tests/test_views.py:690]: Looking good.  Does this test check if we land on a good page somewhere?  i.e. eventually we get a 200 response?

**Review by tylerecouture — CHANGES_REQUESTED** [review]
Looking good, couple minor changes, and a bigger issue is you need to rebase your code.   The upstream will be changing and you need to keep your branch up to date with upstream.  You might need to look into best practices when contributing to a collaborative code base with git.

File: src/courses/tests/test_views.py (hunk 4 above)
Comment (jefcolbi) [reply src/courses/tests/test_views.py:690]: added

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -16,7 +16,8 @@ If you're interested in contributing to this repo, please work through these ste
 5. Create a new branch with a name specific to the issue or feature or bug you will be working on: `git checkout -b yourbranchname`
 6. Write code!
 7. Before committing, make sure to run tests and linting locally (this will save you the annoyance of having to clean up lots of little "oops typo!" commits).  Note that the `--failfast` and `--parallel` modes are optional and used to speed up the tests.  `--failfast` will quit as soon as one test fails, and `--parallel` will run tests in multiple processes (however if a test fails, the output might not be helpful, and you might need to run the tests again without this option to get more info on the failing test):   
-`./src/manage.py test src --failfast --parallel && flake8 src`
+   * venv: `./src/manage.py test src --failfast --parallel && flake8 src`
+   * docker: `docker-compose exec web bash -c "./src/manage.py test src --failfast --parallel && flake8 src"`
```
[the remark this answers is not in the document]
Comment (jefcolbi) [reply CONTRIBUTING.md:20]: it is rebase that caused all those problems. i usually don't use --rebase when i merge

**Review by tylerecouture — APPROVED** [review]
For future reference, you should avoid reformatting using your IDEs default settings when working on a project that isn't your own, lots of reformatting here that doesn't match our repo settings that I will now need to revert.  Try to be more precise in what code you commit.
